core/src/main/java/org/infinispan/factories/ComponentRegistry.java

```diff
@@ -217,6 +221,29 @@ public void start() {
       super.start();
    }
 
+   @Override
+   protected CompletionStage<Void> await() {
+      if (configuration.simpleCache() || icr.running().isInternalCache(cacheName)) return super.await();
+
+      synchronized (this) {
+         if (state == ComponentStatus.INITIALIZING) {
+            if (waitStableTopology() && globalComponents.getLocalTopologyManager() != null) {
+               CompletionStage<Void> cs = globalComponents.getLocalTopologyManager().awaitStableTopology(cacheName);
+               if (cs != null)
+                  return cs;
+            }
+         }
+      }
+
+      return super.await();
+   }
+
+   private boolean waitStableTopology() {
+      // Cache with non-shared stores without purgeOnStart set must wait until a stable topology is installed.
```

Comment:
Why is this only affecting only stores that aren't shared with persistent state? We should be waiting for the stable topology under any configuration. Otherwise we could have a sub cluster form that doesn't have all the values yet and when the remaining member joins it will not be able to bring along its values.